File: 165.compare-version-numbers.py
#
# @lc app=leetcode id=165 lang=python3
#
# [165] Compare Version Numbers
#

# Time:  O(1)
# Space: O(1)

# @lc code=start
class Solution:
    def compareVersion(self, version1: str, version2: str) -> int:
        i, j, n1, n2 = 0, 0, len(version1), len(version2)
        while i < n1 or j < n2:
            v1, v2 = 0, 0
            while i < n1 and version1[i] != ".":
                v1 = v1 * 10 + int(version1[i])
                i += 1
            while j < n2 and version2[j] != ".":
                v2 = v2 * 10 + int(version2[j])
                j += 1
            if v1 != v2:
                return 1 if v1 > v2 else -1
            i += 1
            j += 1
        return 0

        # less efficient
        v1 = [int(v) for v in version1.split(".")]
        v2 = [int(v) for v in version2.split(".")]
        max_len = max(len(v1), len(v2))
        v1.extend([0] * (max_len - len(v1)))
        v2.extend([0] * (max_len - len(v2)))
        for i in range(max_len):
            if v1[i] != v2[i]:
                return 1 if v1[i] > v2[i] else -1
        return 0


# @lc code=end

File: 165.compare-version-numbers.js
/*
 * @lc app=leetcode id=165 lang=javascript
 *
 * [165] Compare Version Numbers
 */

// @lc code=start
/**
 * @param {string} version1
 * @param {string} version2
 * @return {number}
 */
var compareVersion = function (version1, version2) {
  let i = 0;
  let j = 0;
  while (i < version1.length || j < version2.length) {
    let v1 = 0;
    let v2 = 0;
    while (i < version1.length && version1[i] !== ".") {
      v1 = v1 * 10 + Number(version1[i++]);
    }
    while (j < version2.length && version2[j] !== ".") {
      v2 = v2 * 10 + Number(version2[j++]);
    }
    if (v1 !== v2) {
      return v1 > v2 ? 1 : -1;
    }
    ++i, ++j;
  }
  return 0;
};
// @lc code=end
